src/cargo/util/command_prelude.rs

```diff
@@ -986,6 +992,60 @@ pub fn root_manifest(manifest_path: Option<&Path>, gctx: &GlobalContext) -> Carg
     }
 }
 
+pub fn get_tests_from_manifest() -> CargoResult<Vec<String>> {
+    let manifest = get_manifest_with_table()?;
+    let mut tests = vec![];
+
+    if let Some(tests_value) = manifest.get("test") {
+        if let Some(tests_array) = tests_value.as_array() {
+            for test in tests_array {
+                if let Some(test) = test.as_str() {
+                    tests.push(test.to_string());
+                }
+            }
+        }
+    }
+
+    Ok(tests)
+}
+
+pub fn get_tests_candidates() -> Vec<clap_complete::dynamic::CompletionCandidate> {
+    get_tests_from_manifest()
+        .map(|tests| {
+            tests
+                .into_iter()
+                .map(|test| clap_complete::dynamic::CompletionCandidate::new(test))
+                .collect()
+        })
+        .unwrap_or(vec![])
+}
+
+pub fn get_manifest_with_table() -> CargoResult<toml::Table> {
+    // Exec `cargo locate-project --message-format json` to get the manifest path.
+    let output = std::process::Command::new("cargo")
```

Comment:
Since we are in Cargo, we don't need to launch a new instance to do look ups, we can do it with our own APIs